src/gluonts/mx/trainer/callback.py

```diff
@@ -0,0 +1,627 @@
+        for callback in self.callbacks:
+            callback.on_train_epoch_start(training_network=training_network)
+
+    def on_validation_epoch_start(
+        self, training_network: nn.HybridBlock
+    ) -> None:
+        for callback in self.callbacks:
+            callback.on_validation_epoch_start(
+                training_network=training_network
+            )
+
+    def on_train_batch_end(self, training_network: nn.HybridBlock) -> None:
+        for callback in self.callbacks:
+            callback.on_train_batch_end(training_network=training_network)
+
+    def on_validation_batch_end(
+        self, training_network: nn.HybridBlock
+    ) -> None:
+        for callback in self.callbacks:
+            callback.on_validation_batch_end(training_network=training_network)
+
+    def on_train_epoch_end(
+        self,
+        epoch_no: int,
+        epoch_loss: float,
+        training_network: nn.HybridBlock,
+        trainer: gluon.Trainer,
+    ) -> bool:
+        return np.all(
+            [
+                callback.on_train_epoch_end(
+                    epoch_no=epoch_no,
+                    epoch_loss=epoch_loss,
+                    training_network=training_network,
+                    trainer=trainer,
+                )
+                for callback in self.callbacks
+            ]
+        )
+
+    def on_validation_epoch_end(
+        self,
+        epoch_no: int,
+        epoch_loss: float,
+        training_network: nn.HybridBlock,
+        trainer: gluon.Trainer,
+    ) -> bool:
+        return np.all(
+            [
+                callback.on_validation_epoch_end(
+                    epoch_no=epoch_no,
+                    epoch_loss=epoch_loss,
+                    training_network=training_network,
+                    trainer=trainer,
+                )
+                for callback in self.callbacks
+            ]
+        )
+
+    def on_epoch_end(
+        self,
+        epoch_no: int,
+        epoch_loss: float,
+        training_network: nn.HybridBlock,
+        trainer: gluon.Trainer,
+        best_epoch_info: Dict[str, Any],
+        ctx: mx.Context,
+    ) -> bool:
+        return np.all(
+            [
+                callback.on_epoch_end(
+                    epoch_no=epoch_no,
+                    epoch_loss=epoch_loss,
+                    training_network=training_network,
+                    trainer=trainer,
+                    best_epoch_info=best_epoch_info,
+                    ctx=ctx,
+                )
+                for callback in self.callbacks
+            ]
+        )
+
+    def on_train_end(
+        self,
+        training_network: nn.HybridBlock,
+        temporary_dir: str,
+        ctx: mx.context.Context = None,
+    ) -> None:
+        for callback in self.callbacks:
+            callback.on_train_end(
+                training_network=training_network,
+                temporary_dir=temporary_dir,
+                ctx=ctx,
+            )
+
+
```

Comment:
I wonder if we should split the file here to separate the callback definition and implementations.

Reply:
We could add a new folder `callbacks` and add `base.py` and another file per implementation?

Reply:
I'm in favor of keeping all together in one module

Reply:
Should I add the `Callback` base class to `trainer/__init__.py`? It was a bit unintuitive to import when using it.

Reply:
I think we can also add `Callback` directly to `_base.py` since it's so core to the `Trainer` working.